File: 백준/구현/BOJ_15686.py
# 백준 15686 치킨 배달
# 21.01.19
import sys


def getChickenDist(cnt, idx):
    global m, ans
    if cnt == m:
        tempAllDist = 0
        for i in home:
            r1, c1 = i
            tempDist = int(1e9)
            for j in range(len(check)):
                if check[j]:
                    r2, c2 = chicken[j]
                    tempDist = min(tempDist, abs(r1 - r2) + abs(c1 - c2))
            tempAllDist += tempDist
        ans = min(ans, tempAllDist)
        return
    else:
        for k in range(idx, len(chicken)):
            if not check[k]:
                check[k] = True
                getChickenDist(cnt+1, k+1)
                check[k] = False
        return


n, m = map(int, input().split())

graph = [[] for _ in range(n)]

for i in range(n):
    graph[i] = list(map(int, sys.stdin.readline().rstrip().split()))

home = []
chicken = []
ans = 2147483647

for i in range(n):
    for j in range(n):
        if graph[i][j] == 1:
            home.append((i, j))
        elif graph[i][j] == 2:
            chicken.append((i, j))

check = [False] * (len(chicken))

getChickenDist(0, 0)
print(ans)

File: 백준/구현/BOJ_15686.cpp
/*----------------------
|백준 15686 치킨 배달
|21.02.01
----------------------*/

#include <iostream>
#include <algorithm>
#include <queue>
#include <vector>
#include <string>
#include <cmath>
#define INF 2147000000

using namespace std;

int n, m;
vector <pair <int, int> > homes;
vector <pair <bool, pair <int, int> > > chickens;
int ans = INF;

void getDistance(int cnt, int idx)
{
    if (cnt == m)
    {
        int totalDist = 0;
        for (int i = 0; i < homes.size(); ++i)
        {
            int dist = INF;
            for (int j = 0; j < chickens.size(); ++j)
            {
                if (chickens[j].first)
                {
                    int tempDist = abs(homes[i].first - chickens[j].second.first) + abs(homes[i].second- chickens[j].second.second);
                    dist = min(dist, tempDist);
                }
            }
            totalDist += dist;
        }
        ans = min(ans, totalDist);
    }

    for (int i = idx; i < chickens.size(); ++i)
    {
        if (!chickens[i].first)
        {
            chickens[i].first = true;
            getDistance(cnt + 1, i + 1);
            chickens[i].first = false;
        }
    }
}

int main()
{
    scanf("%d %d", &n, &m);

    for (int i = 0; i < n; ++i)
    {
        for (int j = 0; j < n; ++j)
        {
            int temp; scanf("%d", &temp);
            if (temp == 1) homes.push_back({i, j});
            else if (temp == 2) chickens.push_back({false, {i, j}});
        }
    }

    getDistance(0, 0);

    printf("%d\n", ans);

    return 0;
}